File: lib/models/protocol.dart
import 'dart:typed_data';

/// 协议命令常量
class Cmd {
  // Host → ESP8266
  static const int handshake = 0x01;
  static const int format = 0x02;
  static const int play = 0x03;
  static const int stop = 0x04;
  static const int data = 0x05;
  static const int volume = 0x06;
  static const int ping = 0x07;
  static const int reset = 0x08;

  // ESP8266 → Host
  static const int ack = 0x80;
  static const int nack = 0x81;
  static const int ready = 0x82;
  static const int error = 0x83;
}

const int sync1 = 0xAA;
const int sync2 = 0x55;
const int maxPayload = 1024;

/// CRC8 (Dallas/Maxim, 多项式 0x31 反射)
int crc8(Uint8List data) {
  int crc = 0x00;
  for (final byte in data) {
    int b = byte;
    for (int i = 0; i < 8; i++) {
      final mix = (crc ^ b) & 0x01;
      crc >>= 1;
      if (mix != 0) crc ^= 0x8C;
      b >>= 1;
    }
  }
  return crc;
}

/// 构建协议帧
Uint8List buildFrame(int cmd, Uint8List payload) {
  final len = payload.length;
  assert(len <= maxPayload, 'payload too large');

  final frame = Uint8List(5 + len + 1);
  frame[0] = sync1;
  frame[1] = sync2;
  frame[2] = cmd;
  frame[3] = (len >> 8) & 0xFF;
  frame[4] = len & 0xFF;
  frame.setAll(5, payload);

  // CRC8 over CMD+LEN+PAYLOAD
  final crcInput = Uint8List(3 + len);
  crcInput[0] = cmd;
  crcInput[1] = frame[3];
  crcInput[2] = frame[4];
  crcInput.setAll(3, payload);
  frame[5 + len] = crc8(crcInput);

  return frame;
}

/// 构建简单命令帧 (无载荷)
Uint8List buildCmdFrame(int cmd) => buildFrame(cmd, Uint8List(0));

/// 构建 DATA 帧
Uint8List buildDataFrame(int seq, Uint8List samples) {
  final payload = Uint8List(2 + samples.length);
  payload[0] = seq & 0xFF;
  payload[1] = (seq >> 8) & 0xFF;
  payload.setAll(2, samples);
  return buildFrame(Cmd.data, payload);
}

/// 构建 FORMAT 帧
Uint8List buildFormatFrame(int sampleRate, int bitsPerSample, int channels) {
  final payload = Uint8List(6);
  payload[0] = sampleRate & 0xFF;
  payload[1] = (sampleRate >> 8) & 0xFF;
  payload[2] = (sampleRate >> 16) & 0xFF;
  payload[3] = (sampleRate >> 24) & 0xFF;
  payload[4] = bitsPerSample;
  payload[5] = channels;
  return buildFrame(Cmd.format, payload);
}

/// 构建 VOLUME 帧
Uint8List buildVolumeFrame(int level) {
  return buildFrame(Cmd.volume, Uint8List.fromList([level]));
}

/// 解析帧，返回 (cmd, payload) 或 null
(int, Uint8List)? parseFrame(Uint8List frame) {
  if (frame.length < 6) return null;
  if (frame[0] != sync1 || frame[1] != sync2) return null;

  final cmd = frame[2];
  final len = ((frame[3] << 8) | frame[4]) as int;

  if (frame.length != 5 + len + 1) return null;

  final payload = frame.sublist(5, 5 + len);
  final expectedCrc = crc8(frame.sublist(2, 5 + len));
  if (frame[5 + len] != expectedCrc) return null;

  return (cmd, payload);
}

File: rust/src/protocol.rs
/// 协议命令常量
pub mod cmd {
    // 上行 (Host → ESP8266)
    pub const HANDSHAKE: u8 = 0x01;
    pub const FORMAT: u8 = 0x02;
    pub const PLAY: u8 = 0x03;
    pub const STOP: u8 = 0x04;
    pub const DATA: u8 = 0x05;
    pub const VOLUME: u8 = 0x06;
    pub const PING: u8 = 0x07;
    pub const RESET: u8 = 0x08;

    // 下行 (ESP8266 → Host)
    pub const ACK: u8 = 0x80;
    pub const NACK: u8 = 0x81;
    pub const READY: u8 = 0x82;
    pub const ERROR: u8 = 0x83;
}

pub const SYNC1: u8 = 0xAA;
pub const SYNC2: u8 = 0x55;
pub const HEADER_SIZE: usize = 5;
pub const CRC_SIZE: usize = 1;
pub const MAX_PAYLOAD: usize = 1024;

/// CRC8 (Dallas/Maxim, 多项式 0x31 反射)
pub fn crc8(data: &[u8]) -> u8 {
    let mut crc: u8 = 0x00;
    for &byte in data {
        let mut b = byte;
        for _ in 0..8 {
            let mix = (crc ^ b) & 0x01;
            crc >>= 1;
            if mix != 0 {
                crc ^= 0x8C;
            }
            b >>= 1;
        }
    }
    crc
}

/// 构建协议帧，返回完整字节数组
pub fn build_frame(cmd: u8, payload: &[u8]) -> Vec<u8> {
    let len = payload.len();
    assert!(len <= MAX_PAYLOAD, "payload too large");

    let total = HEADER_SIZE + len + CRC_SIZE;
    let mut frame = Vec::with_capacity(total);

    frame.push(SYNC1);
    frame.push(SYNC2);
    frame.push(cmd);
    frame.push((len >> 8) as u8);
    frame.push((len & 0xFF) as u8);
    frame.extend_from_slice(payload);

    let crc = crc8(&frame[2..]); // 计算 CRC8 over CMD + LEN + PAYLOAD
    frame.push(crc);

    frame
}

/// 构建 DATA 帧 (命令 0x05, 载荷 = seq(2字节 LE) + samples)
pub fn build_data_frame(seq: u16, samples: &[u8]) -> Vec<u8> {
    let mut payload = Vec::with_capacity(2 + samples.len());
    payload.push((seq & 0xFF) as u8);
    payload.push((seq >> 8) as u8);
    payload.extend_from_slice(samples);
    build_frame(cmd::DATA, &payload)
}

/// 构建简单命令帧 (无载荷)
pub fn build_cmd_frame(cmd: u8) -> Vec<u8> {
    build_frame(cmd, &[])
}

/// 构建 FORMAT 帧
pub fn build_format_frame(sample_rate: u32, bits_per_sample: u8, channels: u8) -> Vec<u8> {
    let mut payload = Vec::with_capacity(6);
    payload.extend_from_slice(&sample_rate.to_le_bytes());
    payload.push(bits_per_sample);
    payload.push(channels);
    build_frame(cmd::FORMAT, &payload)
}

/// 构建 VOLUME 帧
pub fn build_volume_frame(level: u8) -> Vec<u8> {
    build_frame(cmd::VOLUME, &[level])
}

/// 解析帧，返回 (命令, 载荷)
/// 返回 None 表示帧不完整或校验失败
pub fn parse_frame(frame: &[u8]) -> Option<(u8, Vec<u8>)> {
    if frame.len() < HEADER_SIZE + CRC_SIZE {
        return None;
    }
    if frame[0] != SYNC1 || frame[1] != SYNC2 {
        return None;
    }

    let cmd = frame[2];
    let len = ((frame[3] as usize) << 8) | (frame[4] as usize);

    if frame.len() != HEADER_SIZE + len + CRC_SIZE {
        return None;
    }

    let payload = frame[5..5 + len].to_vec();
    let expected_crc = crc8(&frame[2..5 + len]);
    if frame[5 + len] != expected_crc {
        return None;
    }

    Some((cmd, payload))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_crc8() {
        assert_eq!(crc8(&[0x01, 0x00, 0x00]), 0x4B);
    }

    #[test]
    fn test_build_and_parse_cmd_frame() {
        let frame = build_cmd_frame(cmd::PING);
        let (cmd, payload) = parse_frame(&frame).unwrap();
        assert_eq!(cmd, cmd::PING);
        assert!(payload.is_empty());
    }

    #[test]
    fn test_build_and_parse_data_frame() {
        let samples: Vec<u8> = (0..254).map(|i| (i % 256) as u8).collect();
        let frame = build_data_frame(42, &samples);
        let (cmd, payload) = parse_frame(&frame).unwrap();
        assert_eq!(cmd, cmd::DATA);
        assert_eq!(payload.len(), 2 + 254);
        assert_eq!(payload[0], 42 & 0xFF);
        assert_eq!(payload[1], 42 >> 8);
    }

    #[test]
    fn test_corrupted_frame() {
        let mut frame = build_cmd_frame(cmd::PLAY);
        frame[3] = 0xFF; // corrupt length
        assert!(parse_frame(&frame).is_none());
    }
}
